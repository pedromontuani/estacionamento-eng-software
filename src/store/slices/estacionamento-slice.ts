import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {IPatio, ICliente, IUsuario} from '../../types';
import {initialState} from '../../constants';

export interface IEstacionamentoState {
  patios: IPatio[];
  clientes: ICliente[];
  usuarios: IUsuario[];
  rendimento: number;
  usuarioLogado: IUsuario | null;
}

interface IAddClienteAction {
  idPatio: string;
  cliente: ICliente;
}

const beneficiosSlice = createSlice({
  name: 'estacionamento',
  initialState,
  reducers: {
    ADD_PATIO: (state, {payload}: PayloadAction<IPatio>) => {
      return {...state, patios: [...state.patios, payload]};
    },
    REMOVE_PATIO: (state, {payload}: PayloadAction<IPatio>) => {
      return {
        ...state,
        patios: state.patios.filter(p => p.id !== payload.id),
        clientes: state.clientes.filter(c => c.idPatio !== payload.id),
      };
    },
    ADD_CLIENTE: (state, {payload}: PayloadAction<IAddClienteAction>) => {
      const clientes = [...state.clientes, payload.cliente];
      const patios = state.patios.map(patio => {
        if (patio.id == payload.idPatio) {
          return {
            ...patio,
            vagasOcupadas: patio.vagasOcupadas + 1,
          };
        }
        return patio;
      });
      return {...state, patios, clientes};
    },
    REMOVE_CLIENTE: (state, {payload}: PayloadAction<ICliente>) => {
      const clientes = state.clientes.filter(c => c.cpf !== payload.cpf);
      let rendimento = 0;
      const patios = state.patios.map(patio => {
        if (patio.id == payload.idPatio) {
          rendimento = patio.valorHora;
          return {
            ...patio,
            vagasOcupadas: patio.vagasOcupadas - 1,
          };
        }
        return patio;
      });
      return {
        ...state,
        patios,
        clientes,
        rendimento: state.rendimento + rendimento,
      };
    },
    ADD_USUARIO: (state, {payload}: PayloadAction<IUsuario>) => {
      return {...state, usuarios: [...state.usuarios, payload]};
    },
    REMOVE_USUARIO: (state, {payload}: PayloadAction<IUsuario>) => {
      return {
        ...state,
        usuarios: state.usuarios.filter(u => u.email !== payload.email),
      };
    },
    SIGN_IN: (state, {payload}: PayloadAction<any>) => {
      return {...state, usuarioLogado: payload};
    },
    SIGN_OUT: state => {
      return {...state, usuarioLogado: null};
    },
  },
});

export const {actions, reducer} = beneficiosSlice;

export const {
  ADD_PATIO,
  REMOVE_PATIO,
  ADD_CLIENTE,
  REMOVE_CLIENTE,
  ADD_USUARIO,
  REMOVE_USUARIO,
  SIGN_OUT,
  SIGN_IN,
} = actions;

export const getRendimentosMensal = (state: IEstacionamentoState) => {
  return {
    labels: ['Janeiro', 'Fevereiro', 'MarÃ§o'],
    datasets: [
      {
        data: [0, 0, state.rendimento],
      },
    ],
  };
};

export const getRendimentosSemanal = (state: IEstacionamentoState) => {
  return {
    labels: ['01 - 08', '09 - 16', '17 - 22', '23 - 31'],
    datasets: [
      {
        data: [0, 0, 0, state.rendimento],
      },
    ],
  };
};

export const getAllUsuarios = (state: IEstacionamentoState) => state.usuarios;

export const findUsuarioByEmailAndPassoword = (
  usuarios: IUsuario[],
  email: string,
  password: string,
) => {
  return usuarios.find(u => u.email == email && u.senha == password);
};

export const getPatios = (state: IEstacionamentoState) => state.patios;

export const getPatioById = (state: IEstacionamentoState, idPatio: string) =>
  state.patios.find(p => p.id == idPatio);

export const getCarrosByIdPatio = (
  state: IEstacionamentoState,
  idPatio: string,
) => state.clientes.filter(cliente => cliente.idPatio == idPatio);

export const getUsuarios = (state: IEstacionamentoState) => {
  return state.usuarios.filter(u => u.email !== state.usuarioLogado?.email);
};
