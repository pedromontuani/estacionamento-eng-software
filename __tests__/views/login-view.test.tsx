import React from 'react';
import {fireEvent} from '@testing-library/react-native';

import {rendererCreate, customRender} from '../utils';
import * as estacionamentoSlice from '../../src/store/slices/estacionamento-slice';
import LoginView from '../../src/views/login-view';

describe('Testing login view', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should match snapshot', async () => {
    const tree = rendererCreate(<LoginView />);
    expect(tree).toMatchSnapshot();
  });

  test('should sign in', async () => {
    const {findByTestId} = customRender(<LoginView />);
    const emailInput = await findByTestId('email-input');
    const passwordInput = await findByTestId('password-input');
    const signInButton = await findByTestId('sign-in-button');

    const signIn = jest.spyOn(estacionamentoSlice, 'SIGN_IN');

    await fireEvent.changeText(emailInput, 'admin@estacionamento.com');
    await fireEvent.changeText(passwordInput, '123456');
    await fireEvent.press(signInButton);

    expect(signIn).toHaveBeenCalled();
  });

  test('should not sign in', async () => {
    const {findByTestId} = customRender(<LoginView />);
    const emailInput = await findByTestId('email-input');
    const passwordInput = await findByTestId('password-input');
    const signInButton = await findByTestId('sign-in-button');

    const signIn = jest.spyOn(estacionamentoSlice, 'SIGN_IN');

    await fireEvent.changeText(emailInput, 'teste@estacionamento.com');
    await fireEvent.changeText(passwordInput, '123456');
    await fireEvent.press(signInButton);
    expect(signIn).not.toHaveBeenCalled();
  });
});
