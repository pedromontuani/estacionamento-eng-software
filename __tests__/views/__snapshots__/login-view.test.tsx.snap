// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing login view should match snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": 1334,
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "stretch",
        "backgroundColor": "white",
        "borderRadius": 12,
        "elevation": 5,
        "height": 400.2,
        "justifyContent": "space-evenly",
        "marginBottom": 180,
        "padding": 16,
        "width": 562.5,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "stretch",
          "backgroundColor": "white",
          "borderRadius": 22,
          "elevation": 2,
          "height": 44,
          "justifyContent": "center",
          "paddingHorizontal": 8,
          "width": "100%",
        }
      }
    >
      <TextInput
        autoCapitalize="none"
        keyboardType="email-address"
        onChangeText={[Function]}
        placeholder="Email"
        testID="email-input"
        textContentType="emailAddress"
        value=""
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "stretch",
          "backgroundColor": "white",
          "borderRadius": 22,
          "elevation": 2,
          "height": 44,
          "justifyContent": "center",
          "paddingHorizontal": 8,
          "width": "100%",
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Senha"
        secureTextEntry={true}
        testID="password-input"
        value=""
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#1976d2",
          "borderRadius": 20,
          "elevation": 5,
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": "100%",
        }
      }
      testID="sign-in-button"
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 14,
            "fontWeight": "500",
          }
        }
      >
        LOGIN
      </Text>
    </View>
  </View>
</View>
`;
